pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -799,8 +799,15 @@ protected static Iterator<RecordInfo> resolveComparisonTies(Iterator<RecordInfo>
               return recordInfo;
             }
             int comparisonResult = newComparisonValue.compareTo(maxComparisonValueRecordInfo.getComparisonValue());
-            if (comparisonResult >= 0) {
+            if (comparisonResult > 0) {
               return recordInfo;
+            } else if (comparisonResult == 0) {
+              // Check if mappings are empty or the current record has a higher doc ID position
```

Comment:
> What happens in this scenario: say we have two replicas of a consuming segment R0 and R1. Then we initiate a commit flow:
R0 is selected as the committer and R1 is told to wait.
R0 sorts the records and does a local addOrReplaceSegment. Say the sort order for the records is not unique.
R1 is told to discard its local copy and download from peer.
Would the addOrReplaceSegment call in that case be made with R1's old MutableSegment and R0's ImmutableSegment? If so, we can't use this mapping from the Mutable Segment because the two replicas could have come to different orderings.

In this scenario we would use R0's ImmutableSegment to replace. In this case, we wouldn't have the original docID mapping anymore. If you think about it, we would hit the scenario during other similar events like resetting segment. In order to resolve it properly, we need to persist the original docID order somewhere like segment-metadata or a virtual column. Having a virtual column like `$originalDocID` would also make sense and especially for upserts we can refer that to resolve comparison ties.

> Even for Vanilla Realtime tables we can end up with CRC mismatch due to comparison ties.

This shouldn't be an issue for append-only tables as we are not concerned about data-correctness. In case of upserts, data-correctness becomes a major issue.

> I think the simplest recommendation we should share with customers for now is: ensure that you don't have comparison column ties.

A lot of folks use constant comparison column to ensure upsert happens for every new record for a given key. Maybe we can try to brainstorm possible solutions and see how involved the solution actually is and accordingly make a call.

@Jackie-Jiang any suggestions how we can support this ^